Adding redis session storage

tahmidefaz opened the pull request:
Adding the redis session storage.

josejulio commented on libs/common/src/common/config/app.py:
```diff
@@ -125,10 +125,12 @@
 
 lock_store_type = _config("LOCK_STORE_TYPE", default="in_memory")
 
+use_redis = _config("USE_REDIS", default=False, cast=bool)
```
We could change this to `session_storage` with options `file` or `redis`  - that way we can add more options in the future if needed - and this doesn't add too much overhead :sweat_smile:

tahmidefaz replied:
Sounds good, done! :+1:

josejulio commented on libs/common/src/common/config/app.py:
```diff
@@ -125,10 +125,12 @@
+redis_port = _config("REDIS_PORT", default=6379, cast=int)
 redis_username = _config("REDIS_USERNAME", default=None)
 redis_password = _config("REDIS_PASSWORD", default=None)
+redis_db = _config("REDIS_DB", default=0, cast=int)
```
I suppose this should be a string?

tahmidefaz replied:
I think it can be both, but I changed it to a `str` eitherways. :+1: 
[Here](https://github.com/redis/redis-py/blob/master/redis/asyncio/client.py#L209)

josejulio replied:
Sounds fair, I though this was actually the host

tahmidefaz commented on libs/common/src/common/session_storage/redis.py:
```diff
@@ -0,0 +1,42 @@
+    async def retrieve(self, session_key: str) -> Optional[Session]:    
+        """Read the identity header from Redis using the session_id"""
+        try:
+            value = self.redis_client.get(session_key)
```
Yep! Updated with `redis.asyncio`.

tahmidefaz commented:
@josejulio updated! I had to remove the `jwt` package. There's some dependency collision with the ibm watson libraries, and it was it not letting me spin up the va backend api.

josejulio commented on libs/common/src/common/session_storage/redis.py:
```diff
@@ -0,0 +1,41 @@
+            else:
+                return None
+        except Exception as e:
+            print(f"Error reading from Redis: {e}")
```
Lets allow the error to propagate and be free! 
If this fails, we shouldn't keep processing the request anyway.
I don't think we need to tell the user about something in particular other than "whoops, this failed"

josejulio commented on libs/common/src/common/session_storage/redis.py:
```diff
@@ -0,0 +1,41 @@
+        try:
+            await self.redis_client.set(session.key, session.value)
+            print(f"Session ({session.key} written to Redis.")
+        except Exception as e:
```
Lets allow the error to propagate and be free!

josejulio commented on pyproject.toml:
```diff
@@ -9,6 +9,7 @@ dependencies = [
     "quart-injector==0.0.0b1",
     "pydantic>=2.10.5",
     "logstash_formatter>=0.5.17",
+    "redis>=5.2.1",
```
We shouldn't need this one, we are going to inherit from common in the other projects.

josejulio commented on services/virtual-assistant/src/virtual-assistant/run.py:
```diff
@@ -19,7 +27,21 @@ async def handle_request_validation_error(error):
+
+    # This gets injected into routes when it is requested.
+    # e.g. async def status(session_storage: injector.Inject[SessionStorage]) -> TalkResponse:
+    if app.session_storage == "redis":
+        binder.bind(SessionStorage, to=RedisSessionStorage())
+
+    if app.session_storage == "file":
+        binder.bind(SessionStorage, to=FileSessionStorage(".va-session-storage"))
```
It's the same, but it feels grouped and that makes sense since it is sort of a switch/case sort of stuff. It is easier to read.

```suggestion
    if app.session_storage == "redis":
        binder.bind(SessionStorage, to=RedisSessionStorage())
    elif app.session_storage == "file":
        binder.bind(SessionStorage, to=FileSessionStorage(".va-session-storage"))
```

josejulio commented on services/watson-extension/src/watson-extension/run.py:
```diff
@@ -31,8 +32,11 @@ def configure(binder: injector.Binder) -> None:
     # This gets injected into routes when it is requested.
     # e.g. async def status(session_storage: injector.Inject[SessionStorage]) -> StatusResponse:
-    binder.bind(SessionStorage, to=FileSessionStorage(".va-session-storage"))
+    if app_config.session_storage == "redis":
+        binder.bind(SessionStorage, to=RedisSessionStorage())
 
+    if app_config.session_storage == "file":
+        binder.bind(SessionStorage, to=FileSessionStorage(".va-session-storage"))
```
Same as above

```suggestion
    if app_config.session_storage == "redis":
        binder.bind(SessionStorage, to=RedisSessionStorage())
    elif app_config.session_storage == "file":
        binder.bind(SessionStorage, to=FileSessionStorage(".va-session-storage"))
```

josejulio commented:
sorry, I merged the other PR and now this has a conflict :sweat: 
It renamed the python package to use an underscore - the top project folder is still the same (i.e. services/watson-extension/src/watson_extension) so that python doesn't yell at us.

tahmidefaz commented:
@josejulio I removed the try/catch blocks, the errors should now be propagating to the APIs. I guess we can catch these over there? Moved the `redis` dependency under the `common` package.
The merge conflict is a weird one... it never showed up on my machine when I did a rebase. I pushed up my changes a few minutes ago, and github is still processing it apparently. Something is definately off somewhere. :shrug:

josejulio commented:
> The merge conflict is a weird one... it never showed up on my machine when I did a rebase. I pushed up my changes a few minutes ago, and github is still processing it apparently. Something is definately off somewhere. 🤷

I got a notification about this but still can't see it :/

![image](https://github.com/user-attachments/assets/cee35ef5-c3d0-47c5-a1fb-fc9a959cfb9b)

https://github.com/RedHatInsights/astro-virtual-assistant/pull/583/files/1dc4a6660034a21df39159ae68499c226942b1c7..7ce6076f7100efaecdbdf4cb047273e424f87be9

josejulio commented:
Closing bugged PR

![image](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExM3UzNXp5ZmI1ZXdyNndidG5ycHE0bjI5NTRheDNhcG45NjdpNzk5MSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/XvQXEi09zfmcE/giphy.gif)
